rbs/SigsRewriter.cc

```diff
@@ -187,11 +88,82 @@ unique_ptr<parser::NodeVec> signaturesForNode(core::MutableContext ctx, parser::
         }
     }
 
+    methodSignatures.erase(it);
     return signatures;
 }
 
 } // namespace
 
+void SigsRewriter::checkForUnusedComments() {
+    for (const auto &[lineNumber, comments] : methodSignatures) {
+        for (const auto &sig : comments.signatures) {
+            if (auto e = ctx.beginError(sig.typeLoc, core::errors::Rewriter::RBSUnsupported)) {
+                e.setHeader("Unused type annotation. No method def before next annotation");
+                e.addErrorSection(core::ErrorSection(""));
+            }
+        }
+    }
+}
+
+// Triggered once per file. We iterate over the source code and look for RBS comments
+// Each RBS comment is stored in a hash map, methodSignatures. Key is the method name, value is a vector of Comments
+void SigsRewriter::extractComments() {
+    auto lines = absl::StrSplit(ctx.file.data(ctx).source(), '\n');
+    Comments currentComments;
+    uint32_t offset = 0;
+    uint32_t lineNumber = 0;
+
+    for (const auto &line : lines) {
+        lineNumber++;
+        auto [trimmedLine, leadingWhitespaceCount] = stripAsciiWhitespaceWithCount(line);
+
+        // Empty lines between the RBS Comment and the method definition are allowed
+        if (trimmedLine.empty()) {
+            offset += line.length() + 1;
+            continue;
+        }
+
+        if (absl::StartsWith(trimmedLine, RBS_PREFIX)) {
+            auto signature = absl::StripPrefix(trimmedLine, RBS_PREFIX);
+            uint32_t startOffset = offset + leadingWhitespaceCount;
+            uint32_t endOffset = startOffset + trimmedLine.length();
+
+            rbs::Comment comment{
+                core::LocOffsets{startOffset, endOffset},
+                core::LocOffsets{startOffset + 2, endOffset},
+                signature,
+            };
+            currentComments.signatures.emplace_back(move(comment));
+            offset += line.length() + 1;
+            continue;
+        } else if (absl::StartsWith(trimmedLine, ANNOTATION_PREFIX)) {
+            auto annotation = absl::StripPrefix(trimmedLine, ANNOTATION_PREFIX);
```

Comment:
Looks like we scanning twice, once to find if the prefix is here and then once to strip it.

Can't we instead find the index of the prefix and substr index + prefix_size?